src/proposals/proposal.cpp

```diff
@@ -73,32 +116,94 @@ namespace hypha
             // create edge for FAILED_PROPS
             Edge::write(m_dao.get_self(), m_dao.get_self(), root, proposal.getHash(), common::FAILED_PROPS);
         }
-
-        eosio::action(
-            eosio::permission_level{m_dao.get_self(), name("active")},
-            m_dao.getSettingOrFail<eosio::name>(TELOS_DECIDE_CONTRACT), name("closevoting"),
-            std::make_tuple(ballot_id, true))
-            .send();
     }
 
     ContentGroup Proposal::makeSystemGroup(const name &proposer,
                                            const name &proposal_type,
-                                           const string &decide_title,
-                                           const string &decide_desc,
-                                           const string &decide_content)
+                                           const string &decide_title)
     {
         return ContentGroup{
             Content(CONTENT_GROUP_LABEL, SYSTEM),
             Content(CLIENT_VERSION, m_dao.getSettingOrDefault<std::string>(CLIENT_VERSION, DEFAULT_VERSION)),
             Content(CONTRACT_VERSION, m_dao.getSettingOrDefault<std::string>(CONTRACT_VERSION, DEFAULT_VERSION)),
-            Content(BALLOT_ID, registerBallot(proposer, decide_title, decide_desc, decide_content)),
             Content(NODE_LABEL, decide_title),
             Content(TYPE, proposal_type)};
     }
 
-    bool Proposal::didPass(const name &ballot_id)
+    ContentGroup Proposal::makeBallotGroup()
     {
-        name trailContract = m_dao.getSettingOrFail<eosio::name>(TELOS_DECIDE_CONTRACT);
+
+        auto expiration = time_point_sec(current_time_point()) + m_dao.getSettingOrFail<int64_t>(VOTING_DURATION_SEC);
+        return ContentGroup{
+            Content(CONTENT_GROUP_LABEL, BALLOT),
+            Content("expiration", expiration)
+        };
+    }
+
+    ContentGroup Proposal::makeBallotOptionsGroup()
+    {
+        return ContentGroup{
+            Content(CONTENT_GROUP_LABEL, BALLOT_OPTIONS),
+            Content("pass", name("pass")),
+            Content("fail", name("fail"))
+        };
+    }
+
+    void Proposal::updateVoteTally(Document& proposal, const eosio::name creator)
+    {
+        auto [exists, oldTally] = Edge::getIfExists(m_dao.get_self(), proposal.getHash(), common::VOTE_TALLY);
+        if (exists) {
+            oldTally.erase();
+        }
+
+        ContentGroup* contentOptions = proposal.getContentWrapper().getGroupOrFail(BALLOT_OPTIONS);
+
+        std::map<std::string, int64_t> optionsTally;
+        for (auto it = contentOptions->begin(); it != contentOptions->end(); ++it) 
+        {
+            if (it->label != CONTENT_GROUP_LABEL) {
+                optionsTally[it->getAs<std::string>()] = 0;
+            }
+        }
+
+        
+        std::vector<Edge> edges = Edge::getAll(m_dao.get_self(), proposal.getHash(), common::VOTE);
```

Comment:
Just FWIW the same functionality exists on **DocumentGraph** class `m_dao.getEdgesFrom(...)`.

Reply:
Ah... Thanks for sharing!